templates/common/_base/files/configure-ovs-network.yaml

```diff
@@ -0,0 +1,313 @@
+    if ! iface_mac=${ovs_interfaces__mac_address[0]:-$(<"/sys/class/net/${iface}/address")}; then
+      echo "Unable to determine default interface MAC"
+      exit 1
+    fi
+
+    echo "MAC address found for iface: ${iface}: ${iface_mac}"
+
+    # find MTU from config or on original iface
+    iface_mtu=${ovs_interfaces__mtu[0]:-$(ip -j link show "$iface" | jq -r '.[0].mtu')}
+    if [ -z "$iface_mtu" ]; then
+      echo "Unable to determine default interface MTU, defaulting to 1500"
+      iface_mtu=1500
+    else
+      echo "MTU found for iface: ${iface}: ${iface_mtu}"
+    fi
+
+    # create bridge
+    if ! nmcli connection show br-ex &> /dev/null; then
+      nmcli c add type ovs-bridge conn.interface br-ex con-name br-ex 802-3-ethernet.mtu ${iface_mtu} 802-3-ethernet.cloned-mac-address ${iface_mac}
+    fi
+
+    old_conns=()
+    if "$BONDING"; then
+      iface=bond0
+      # store old connections for bringing down later
+      for bond_iface in ${bonded_ifaces[@]}; do
+        old_conns+=($(nmcli --fields UUID,DEVICE conn show --active | grep ${bond_iface} | awk '{print $1}'))
+      done
+    else
+      old_conns+=($(nmcli --fields UUID,DEVICE conn show --active | grep ${iface} | awk '{print $1}'))
+    fi
+
+    # find default port to add to bridge
+    if ! nmcli connection show ovs-port-phys0 &> /dev/null; then
+      nmcli c add type ovs-port conn.interface ${iface} master br-ex con-name ovs-port-phys0 ${phys_port_opts}
+    fi
+
+    if ! nmcli connection show ovs-port-br-ex &> /dev/null; then
+      nmcli c add type ovs-port conn.interface br-ex master br-ex con-name ovs-port-br-ex
+    fi
+
+    # bring down any old iface or bond connection (used for upgrading from local gw to shared gw mode)
+    for old_conn in ${old_conns[@]}; do
+      nmcli conn down $old_conn
+    done
+
+    if "$BONDING"; then
+      # Attach each interface as a slave to the bond ovs port
+      for bond_iface in ${bonded_ifaces[@]}; do
+        if ! nmcli connection show ovs-if-${bond_iface} &> /dev/null; then
+          nmcli c add type 802-3-ethernet conn.interface ${bond_iface} master ovs-port-phys0 con-name ovs-if-${bond_iface} \
+            connection.autoconnect-priority 100 802-3-ethernet.mtu ${iface_mtu}
+        fi
+      done
+    else
+      if ! nmcli connection show ovs-if-phys0 &> /dev/null; then
+        nmcli c add type 802-3-ethernet conn.interface ${iface} master ovs-port-phys0 con-name ovs-if-phys0 \
+          connection.autoconnect-priority 100 802-3-ethernet.mtu ${iface_mtu}
+      fi
+    fi
+
+    if ! nmcli connection show ovs-if-br-ex &> /dev/null; then
+      nmcli c add type ovs-interface slave-type ovs-port conn.interface br-ex master ovs-port-br-ex con-name \
+        ovs-if-br-ex 802-3-ethernet.mtu ${iface_mtu} 802-3-ethernet.cloned-mac-address ${iface_mac}
+    fi
+
+    # wait for DHCP to finish, verify connection is up
+    counter=0
+    while [ $counter -lt 5 ]; do
```

Comment:
Why is this hard-coded to 5 retries 5s apart? What if it takes longer to bring up the interfaces?

Reply:
NM will attempt a certain number of tries to DHCP (it's either 4 or 5) and then give up for 5 minutes. I feel like if we don't have a dhcp response in 25 seconds something is wrong. Then it will try for another 25 seconds on line 100 to bring up the connection with nmcli conn up. This is a blocking operation, so it will not return until NM gives up, and then we will try 5 more times there. tl;dr we are trying for a while here.